prune/service.go

```diff
@@ -0,0 +1,208 @@
+	if len(pruneConfig.GetRemoteFilePrefix()) > 0 {
+		objectName = fmt.Sprintf("%s/%s", pruneConfig.GetRemoteFilePrefix(), objectName)
+	}
+	return objectName
+}
+
+var (
+	initLocalArchiveManager = func(pruneConfig config.MessagePruningConfig) (*ArchiveWriteManager, error) {
+		now := time.Now().UTC().Format("2006_01_02T15_04_05Z")
+		dirPath := fmt.Sprintf("file://%s/%s", pruneConfig.GetExportPath(), pruneConfig.GetRemoteFilePrefix())
+		objectName := fmt.Sprintf("local_%s_%s.jsonl", pruneConfig.GetExportNodeName(), now)
+		log.Info().Msgf("Local archive path: %s, object name: %s", dirPath, objectName)
+		fileBucket, err := blob.OpenBucket(context.Background(), dirPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open local archive file: %w", err)
+		}
+		log.Debug().Msgf("Local archive file opened: %s", dirPath)
+		return NewArchiveWriteManager(NewBlobBucket(fileBucket),
+			objectName, int64(pruneConfig.GetMaxArchiveFileSizeInMB())*1024*1024)
+	}
+
+	initRemoteArchiveManager = func(pruneConfig config.MessagePruningConfig) (*ArchiveWriteManager, error) {
+		if pruneConfig.GetRemoteExportURL() == nil {
+			return nil, nil
+		}
+		objectName := buildRemoteObjectName(pruneConfig)
+		ctx := context.Background()
+		var bucket *blob.Bucket
+		var err error
+		bucket, err = blob.OpenBucket(ctx, pruneConfig.GetRemoteExportURL().String())
+		if err != nil {
+			return nil, fmt.Errorf("failed to open remote bucket: %w", err)
+		}
+		return NewArchiveWriteManager(NewBlobBucket(bucket), objectName, int64(pruneConfig.GetMaxArchiveFileSizeInMB())*1024*1024)
+	}
+
+	initArchiveDirector = func(config config.MessagePruningConfig) (*ArchiveDirector, error) {
+		localArchiveManager, err := initLocalArchiveManager(config)
+		if err != nil {
+			return nil, err
+		}
+		log.Info().Msg("Local archive manager initialized")
+		remoteArchiveManager, err := initRemoteArchiveManager(config)
+		if err != nil {
+			return nil, err
+		}
+		log.Info().Msg("Remote archive manager initialized")
+		return &ArchiveDirector{
+			LocalArchiveManager:  localArchiveManager,
+			RemoteArchiveManager: remoteArchiveManager,
+		}, nil
+	}
+
+	archiveMessage = func(message *data.Message, jobs []*data.DeliveryJob, director *ArchiveDirector) error {
+		archiveData := struct {
+			Message *data.Message       `json:"message"`
+			Jobs    []*data.DeliveryJob `json:"jobs"`
+		}{
+			Message: message,
+			Jobs:    jobs,
+		}
+		jsonData, err := json.Marshal(archiveData)
+		if err != nil {
+			return fmt.Errorf("failed to marshal message and jobs to JSON: %w", err)
+		}
+		jsonStr := string(jsonData) + "\n"
+
+		_, err = director.LocalArchiveManager.Write(context.Background(), jsonStr)
+		if err != nil {
+			return fmt.Errorf("failed to write message and jobs to local archive: %w", err)
+		}
+		if director.RemoteArchiveManager != nil {
+			_, err = director.RemoteArchiveManager.Write(context.Background(), jsonStr)
```

Comment:
Are we writing to remote destination one message at a time? 🤔

Reply:
I would not make any difference if its one at a time or 100s; since the write would be flushed only at `.close()` or when the file reaches rotation size. With 20MB file size we observed approximately 3000 messages are written to a file.